sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala

```diff
@@ -193,14 +228,58 @@ private[sql] class JacksonGenerator(
    *
    * @param row The row to convert
    */
-  def write(row: InternalRow): Unit = writeObject(writeFields(row, schema, rootFieldWriters))
+  def write(row: InternalRow): Unit = dataType match {
+    case st: StructType =>
+      writeObject(writeFields(row, st, rootFieldWriters))
+    case _ => throw new UnsupportedOperationException(
+      s"`JacksonGenerator` can only be used to write out a row when initialized with `StructType`.")
+  }
 
   /**
-   * Transforms multiple `InternalRow`s to JSON array using Jackson
+   * Transforms multiple `InternalRow`s or `MapData`s to JSON array using Jackson
    *
-   * @param array The array of rows to convert
+   * @param array The array of rows or maps to convert
    */
-  def write(array: ArrayData): Unit = writeArray(writeArrayData(array, arrElementWriter))
+  def write(array: ArrayData): Unit = dataType match {
+    case st: StructType =>
+      try {
```

Comment:
If we assume that the caller know what it's doing, we can get rid of this check.

Reply:
I think you're right. If the caller has an error calling, it will also throw a `ClassCastException` to remind the caller. So we're unnecessary to check it.
I'll remove it. Thanks.